_posts/2022-08-22-fan-out-and-back-in-using-functions.md

```diff
@@ -0,0 +1,471 @@
+The fan-in pattern can be applied if you need to wait for all sub tasks to complete before moving on to the next processing step. It can be used to collect and combine the result from each individual sub task.
+
+To fan-in, some kind of shared storage, like a database, a Redis key or an S3 bucket is required for tracking progress of the individual sub tasks.
+
+![fan-in conceptual diagram](/images/2022-fan-out-and-back-in-using-functions/fan-in.png)
+> Diagram of fan-in with functions
+
+## Implement a fan-out/fan-in pattern in OpenFaaS.
+In this section we are going to show you how this pattern can be implemented with OpenFaaS through a relatable use-case. A common big data scenario is batch processing of a data set. In this scenario data is retrieved from a storage system and then processed by parallelized jobs. The results of each individual job is persisted in some kind of data store where it can later be retrieved for further processing when the batch is completed.
+
+In this example we are going to process a CSV file containing URLs to images on Wikipedia. For each URL in the data set we want to run the Inception model and get back image categorizations through machine learning. The inception model can classify what's in a photo based upon a training set made up of objects and animals. As a final processing step we will combine all the individual results into a single output.
+
+![Fan-out/fan-in pattern with OpenFaaS functions](/images/2022-fan-out-and-back-in-using-functions/fan-out-in-example.png)
+> Fan-out/fan-in pattern with OpenFaaS functions
+
+You can find the [full example on GitHub](https://github.com/welteki/openfaas-fan-in-example)
+
+### Fan-out
+S3 will be used as the data store throughout this example. The input files are retrieved from an S3 bucket, results of invocations and the final result will be uploaded to the same bucket. To run this example yourself you will need to create an [Amazon S3 bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/GetStartedWithS3.html).
+
+The functions need access to AWS credentials in order to use the bucket. The `faas-cli` can be used to define these secrets.
+```bash
+echo $aws_access_key_id | faas-cli secret create s3-key
+echo $aws_secret_access_key | faas-cli secret create s3-secret
+```
+
+> You can checkout the documentation for more info on [how to use secrets within your functions](https://docs.openfaas.com/reference/secrets/).
+
+The first part of the workflow will consist of two functions. The `creat-batch` function is responsible for initializing the workflow. It accepts the name of a CSV file stored in an S3 bucket as input. The function will retrieve the CSV file containing image URLs and invoke the second function, `run-model` for each URL in the file.
+
+The `run-model` function is responsible for calling the machine learning model, in this case the `inception` function, and uploading the result to S3.
+
+We will use the `python3-http` template for both functions:
```

Comment:
might be worth elaborating slightly here that `python3-http` is available from the `python3-flask` store template.  It would save a new reader from querying whether lines 66 and 70 are typos as they look inconsistent.